app/Http/Controllers/LinkController.php

```diff
@@ -51,35 +51,54 @@ public function performRedirect(Request $request, $short_url, $secret_key=false)
         $link = Link::where('short_url', $short_url)
             ->first();
 
-        // Return 404 if link not found
+        // Return an error if the link does not exist
+        // or return a 404 if SETTING_REDIRECT_404 is set to true
         if ($link == null) {
-        	return abort(404);
+            if (env('SETTING_REDIRECT_404')) {
+                return redirect()->to(env('SETTING_INDEX_REDIRECT'));
+            }
+            
+            return response(view('error', [
+                'message' => 'Sorry, but this link does not exist.'
+            ]), 404);
         }
 
         // Return an error if the link has been disabled
         // or return a 404 if SETTING_REDIRECT_404 is set to true
         if ($link->is_disabled == 1) {
             if (env('SETTING_REDIRECT_404')) {
-                return abort(404);
+                return redirect()->to(env('SETTING_INDEX_REDIRECT'));
             }
 
-            return view('error', [
+            return response(view('error', [
                 'message' => 'Sorry, but this link has been disabled by an administrator.'
-            ]);
+            ]), 404);
         }
 
-        // Return a 403 if the secret key is incorrect
+        // Return a 404 if the secret key is incorrect
         $link_secret_key = $link->secret_key;
         if ($link_secret_key) {
         	if (!$secret_key) {
         		// if we do not receieve a secret key
-        		// when we are expecting one, return a 403
-        		return abort(403);
+        		// when we are expecting one, return a 404 to keep the link secret
+                if (env('SETTING_REDIRECT_404')) {
+                    return redirect()->to(env('SETTING_INDEX_REDIRECT'));
+                }
+                
+                return response(view('error', [
+                    'message' => 'Sorry, but this link does not exist.'
+                ]), 404);
         	}
         	else {
         		if ($link_secret_key != $secret_key) {
-        			// a secret key is provided, but it is incorrect
-        			return abort(403);
+        			// a secret key is provided, but it is incorrect, return a 404 to keep the link secret
+                    if (env('SETTING_REDIRECT_404')) {
+                        return redirect()->to(env('SETTING_INDEX_REDIRECT'));
+                    }
+                    
+                    return response(view('error', [
+                        'message' => 'Sorry, but this link does not exist.'
+                    ]), 404);
```

Comment:
L57-63, L83-90 and L94-101 are very repetitive. Perhaps we can create a function (e.g `linkNotFound`) that performs these tasks for us?

Additionally, since we are performing the tasks previously handled by `handler.php` in this controller, we should clean up `handler.php` to avoid repetition.